designs/NuGetExtensibility/INuGetProjectServices.md

```diff
@@ -0,0 +1,122 @@
+
+## Problem Background
+
+The existing IVsPackageInstallerServices service is synchronous, and is being called by several components on the UI thread, causing UI delays in Visual Studio. We are creating new async APIs that will allow us to unblock the UI thread, and use the Visual Studio Service Broker to work in Visual Studio CodeSpaces (and maybe Live Share).
+
+See [NuGet extensibility services](NuGetExtensibilityServices.md) for more details.
+
+## Who are the customers
+
+Visual Studio component and extension developers.
+
+## Goals
+
+* Can be called on any thread (ideally background thread, but work if called on UI thread).
+* Quickly return incomplete task to unblock thread.
+* Works with Visual Studio Codespaces, and hopefully Live Share.
+
+## Non-Goals
+
+* This will not be a 1:1 replacement for any of NuGet's existing `IVs` interfaces.
+
+## Solution
+
+```cs
+/// <summary>Service to interact with projects in a solution</summary>
+public interface INuGetProjectServices
+{
+    /// <summary>Gets the list of packages installed in a project.</summmry>
+    /// <param name="project">Path and filename to project file.</param>
+    /// <param name="options">Options to use. If null, uses GetInstalledPakcagesOptions.Default.</param>
+    /// <param name="cancellationToken">Cancellation token.</param>
+    /// <returns>
+    /// The list of packages in the project.
+    /// If the project does not exist in the solution, or the project is unloaded, throws ArgumentException.
+    /// </return>
+    Task<GetInstalledPackagesResult> GetInstalledPackagesAsync(string project, CancellationToken cancellationToken);
+}
+
+/// <summary>Result of a call to INuGetProjectServices.GetInstalledPackagesAsync</summary>
+public sealed class GetInstalledPackagesResult
+{
+    /// <summary>The status of the result</summary>
+    public GetInstalledPackageResultStatus Status { get; }
+
+    /// <summary>List of packages in the project</summary>
+    /// <remarks></remarks>
+    public IReadOnlyCollection<InstalledPackage> Packages { get; }
+}
+
+/// <summary>The status of the result</summary>
+public enum  GetInstalledPackageResultStatus
+{
+    /// <summary>Successful</summary>
+    Successful,
+
+    /// <summary>The project is not yet ready</summary>
+    /// <remarks>This typically happens shortly after the project is loaded, but the project system has not yet informed NuGet about package references</remarks>
+    ProjectNotReady,
+
+    /// <summary>Package information could not be retrieved because the project is in an invalid state</summary>
+    /// <remarks>If a project has an invalid target framework value, or a package reference has a version value, NuGet may be unable to generate basic project information, such as requested packages.</remarks>
+    ProjectInvalid
+}
+
+/// <summary>Basic information about a package</summary>
+public sealed class InstalledPackage
+{
+    /// <summary>The package id.</summary>
+    public string Id { get; }
```

Comment:
Why not [`PackageIdentity`](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Packaging/Core/PackageIdentity.cs) instead o defining `Id` and `Version` ?

Reply:
`PackageIdentity` is not in the BCL, and we've had problem is in the past where other components ship NuGet dlls in Visual Studio. Resharper had to resign NuGet's dlls with a different strong name key to get around this issue. Therefore, we should never use types defined in any NuGet assembly in our VS extensibility APIs.